### PR TITLE
Nested ROI

### PR DESCRIPTION
**Pull request by friskluft**

### CODE REVIEW

File: src/nyx/results_cache.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+	void add_string (const std::string& s) { stringColBuf.push_back(s); }
+	void add_numeric(double n) { calcResultBuf.push_back(n); }
+	void inc_num_rows() { totalNumLabels++; }
+	size_t get_num_rows() { return totalNumLabels; }
+
+private:
+
+	std::vector<double> calcResultBuf;
+	size_t totalNumFeatures = 0, totalNumLabels = 0;
```
Comment (sameeul) [inline src/nyx/results_cache.h:41]: are we using totalNumFeatures for anything?

File: src/nyx/image_loader1x.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	size_t get_tile_height();
+	size_t get_tile_width();
+	size_t get_tile_x(size_t pixel_col);
+	size_t get_tile_y(size_t pixel_row);
+	size_t get_within_tile_idx(size_t pixel_row, size_t pixel_col);
+	std::tuple<uint32_t, uint32_t, uint32_t>  get_image_dimensions(const std::string& filePath);
+	std::tuple<uint32_t, uint32_t, uint32_t>  calculate_tile_dimensions(const std::string& filePath);
+	bool checkTileStatus(const std::string& filePath);
+private:
+	fl::AbstractTileLoader<fl::DefaultView<uint32_t>>* FL = nullptr;
```
Comment (sameeul) [inline src/nyx/image_loader1x.h:32]: can we store this as a unique_ptr instead of raw pointer?

File: src/nyx/image_loader1x.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	size_t fh,
+		fw,
+		fd;
+
+	// Number of tiles along width, height, and depth
+	size_t ntw,
+		nth,
+		ntd;
+
+	int lvl = 0,	// Pyramid level
```
Comment (sameeul) [inline src/nyx/image_loader1x.h:52]: perhaps we should use size_t to stay consistent with row and col?

File: src/nyx/phase1.cpp [hunk 4]
```diff
@@ -44,8 +44,9 @@ namespace Nyxus
 			lyr = 0; //	Layer
 
 		// Open an image pair
-		ImageLoader imlo;
-		bool ok = imlo.open (intens_fpath, label_fpath);
+		ImageLoader & imlo = theImLoader;	//???	ImageLoader imlo;
+		bool ok = true;	//???	bool ok = imlo.open (intens_fpath, label_fpath);
```
Comment (sameeul) [inline src/nyx/phase1.cpp:48]: don't we need to call imlo.open(...) ?

**Review by JesseMckinzie — COMMENTED** [review]
Nyxushie should not quit after missing a parent segment but should print an error message and proceed with the remainder of the data. (line 645-647 of nested_roi_py.cpp)

**Comment by hsidky** [comment]
Let's not forget to add info on this to the main README.md and also to the documentation.

File: src/nyx/results_cache.h (hunk 1 above)
Comment (friskluft) [reply src/nyx/results_cache.h:41]: Yes, it's used in forming result data frames in the Python API:
    // new_bindings_py.cpp
    // ...
    auto nRows = theResultsCache.get_num_rows();
    pyStrData = pyStrData.reshape({nRows, pyStrData.size() / nRows});
    pyNumData = pyNumData.reshape({ nRows, pyNumData.size() / nRows });
   // ...

File: src/nyx/image_loader1x.h (hunk 2 above)
Comment (friskluft) [reply src/nyx/image_loader1x.h:32]: Yes we can

File: src/nyx/image_loader1x.h (hunk 3 above)
Comment (friskluft) [reply src/nyx/image_loader1x.h:52]: Fixed

File: src/nyx/phase1.cpp (hunk 4 above)
Comment (friskluft) [reply src/nyx/phase1.cpp:48]: The loader is being open and closed earlier, in scan_fastloader_way.cpp/processDataset(), so there's no need for calling open() here on the same global image loader instance. I cleaned the code in order for it to look less confusing

File: src/nyx/image_loader1x.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+private:
+	fl::AbstractTileLoader<fl::DefaultView<uint32_t>>* FL = nullptr;
+	std::shared_ptr<std::vector<uint32_t>> ptr = nullptr;
+	// Tile height, width, and depth
+	size_t th,
+		tw,
+		td;
+
+	// 2D tile size
+	size_t tileSize;
```
[the remark this answers is not in the document]
Comment (friskluft) [reply src/nyx/image_loader1x.h:40]: Eventually we need to but that would take a separate PR to fix all Nyxus class data members. I fixed this PR's ones.

File: src/nyx/scan_fastloader_way.cpp [hunk 6]
```diff
@@ -22,50 +22,56 @@ namespace Nyxus
...
 		{
-			LR& r = roiData[lab];
-			size_t footprint = r.get_ram_footprint_estimate();
-			if (footprint >= theEnvironment.get_ram_limit())
+			STOPWATCH("Image scan/ImgScan/Scan/lightsteelblue", "\t=");
+
+			// Report the amount of free RAM
+			unsigned long long freeRamAmt = getAvailPhysMemory();
+			static unsigned long long initial_freeRamAmt = 0;
+			if (initial_freeRamAmt == 0)
```
[the remark this answers is not in the document]
Comment (friskluft) [reply src/nyx/scan_fastloader_way.cpp:34]: No, "static" ensures that the value of "initial_freeRamAmt" persists after the control leaves the function.

File: src/nyx/results_cache.h (hunk 1 above)
Comment (sameeul) [reply src/nyx/results_cache.h]: I am not seeing `totalNumFeatures_` being used by any member function. Can you please double-check?

**Review by sameeul — COMMENTED** [review]
Please take a look at those new comments

File: src/nyx/results_cache.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+	void add_string (const std::string& s) { stringColBuf_.push_back(s); }
+	void add_numeric(double n) { calcResultBuf_.push_back(n); }
+	void inc_num_rows() { totalNumLabels_++; }
+	size_t get_num_rows() { return totalNumLabels_; }
+
+private:
+
+	std::vector<double> calcResultBuf_;
+	size_t totalNumFeatures_ = 0, totalNumLabels_ = 0;
```
[the remark this answers is not in the document]
Comment (friskluft) [reply src/nyx/results_cache.h]: I removed this member

File: src/nyx/helpers/helpers.h [hunk 8]
```diff
@@ -183,5 +183,6 @@ namespace Nyxus
 		return new_pi;
 	}
 
+	bool parse_as_float(std::string raw, float& result);
```
[the remark this answers is not in the document]
Comment (friskluft) [reply src/nyx/helpers/helpers.h]: It's defined in file main_nyxushie.cpp

**Pull request merged**